### ISSUE TITLE
[ERROR] Costmap filters launch: InvalidFrontendLaunchFileError 

### ISSUE THREAD

**Issue by jesusramondovale** [ros-planning/navigation_tutorials#23]
O.S : Linux Ubuntu 22.04 LTS (x86)
ROS2 - Humble Hawksbill 

Followed the Keep Out filter tutorial from 
https://docs.nav2.org/tutorials/docs/navigation2_with_keepout_filter.html
and saw this error while launching 

```$ ros2 launch nav2_costmap_filters_demo costmap_filter_info.launch.py params_file:=`pwd`/src/navigation2_tutorials/nav2_costmap_filters_demo/params/keepout_params.yaml mask:=`pwd`/src/navigation2_tutorials/nav2_costmap_filters_demo/maps/keepout_mask.yaml use_composition:=True```

```
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Caught multiple exceptions when trying to load file of format [py]:
 - ImportError: cannot import name 'NotEqualsSubstitution' from 'launch.substitutions' (/opt/ros/humble/lib/python3.10/site-packages/launch/substitutions/__init__.py)
 - InvalidFrontendLaunchFileError: The launch file may have a syntax error, or its format is unknown
```

Made it work doing this changes to python launcher: 

```
import os

from ament_index_python.packages import get_package_share_directory

from launch import LaunchDescription
from launch.actions import DeclareLaunchArgument
from launch.substitutions import LaunchConfiguration
from launch_ros.actions import Node
from nav2_common.launch import RewrittenYaml


def generate_launch_description():
    # Get the launch directory
    costmap_filters_demo_dir = get_package_share_directory('mss_keepout_filter')

    # Create our own temporary YAML files that include substitutions
    lifecycle_nodes = ['filter_mask_server', 'costmap_filter_info_server']

    # Parameters
    namespace = LaunchConfiguration('namespace')
    use_sim_time = LaunchConfiguration('use_sim_time')
    autostart = LaunchConfiguration('autostart')
    params_file = LaunchConfiguration('params_file')
    mask_yaml_file = LaunchConfiguration('mask')

    # Declare the launch arguments
    declare_namespace_cmd = DeclareLaunchArgument(
        'namespace',
        default_value='',
        description='Top-level namespace')

    declare_use_sim_time_cmd = DeclareLaunchArgument(
        'use_sim_time',
        default_value='true',
        description='Use simulation (Gazebo) clock if true')

    declare_autostart_cmd = DeclareLaunchArgument(
        'autostart', default_value='true',
        description='Automatically startup the nav2 stack')

    declare_params_file_cmd = DeclareLaunchArgument(
            'params_file',
            default_value=os.path.join(costmap_filters_demo_dir, 'params', 'keepout_params.yaml'),
            description='Full path to the ROS 2 parameters file to use')

    declare_mask_yaml_file_cmd = DeclareLaunchArgument(
            'mask',
            default_value=('/root/MecaluxFileSystem/Maps/Current/GeneralCostmap/ActiveKeepoutMap.yaml'),
            description='Full path to filter mask yaml file to load')

    # Make re-written yaml
    param_substitutions = {
        'use_sim_time': use_sim_time,
        'yaml_filename': mask_yaml_file}

    configured_params = RewrittenYaml(
        source_file=params_file,
        root_key=namespace,
        param_rewrites=param_substitutions,
        convert_types=True)

    # Nodes launching commands
    start_lifecycle_manager_cmd = Node(
            package='nav2_lifecycle_manager',
            executable='lifecycle_manager',
            name='lifecycle_manager_costmap_filters',
            namespace=namespace,
            output='screen',
            emulate_tty=True,  # https://github.com/ros2/launch/issues/188
            parameters=[{'use_sim_time': use_sim_time},
                        {'autostart': autostart},
                        {'node_names': lifecycle_nodes}])

    start_map_server_cmd = Node(
            package='nav2_map_server',
            executable='map_server',
            name='filter_mask_server',
            namespace=namespace,
            output='screen',
            emulate_tty=True,  # https://github.com/ros2/launch/issues/188
            parameters=[configured_params])

    start_costmap_filter_info_server_cmd = Node(
            package='nav2_map_server',
            executable='costmap_filter_info_server',
            name='costmap_filter_info_server',
            namespace=namespace,
            output='screen',
            emulate_tty=True,  # https://github.com/ros2/launch/issues/188
            parameters=[configured_params])

    ld = LaunchDescription()

    ld.add_action(declare_namespace_cmd)
    ld.add_action(declare_use_sim_time_cmd)
    ld.add_action(declare_autostart_cmd)
    ld.add_action(declare_params_file_cmd)
    ld.add_action(declare_mask_yaml_file_cmd)

    ld.add_action(start_lifecycle_manager_cmd)
    ld.add_action(start_map_server_cmd)
    ld.add_action(start_costmap_filter_info_server_cmd)

    return ld

```

**Comment by SteveMacenski** [comment]
The tutorials target `rolling` which might have different APIs than Humble. You should be able to fix that easily though with `condition=IfCondition(PythonExpression(['not ', slam])),` instead of `NotEqualsSubstitution` (replace `slam` with the relevant variable causing the issue). The `NotEqualsSubstitution` was introduced in Iron and newer so older distributions don't have it. 

But, removing the conditional as you found also works! Generally speaking composition is very helpful, so I'm highly suggest using it over launching individual nodes where you can.

**Comment by jesusramondovale** [comment]
Hi! I'm suffering another problem with this plugin. Yesterday it worked great but now it's stopping at this point and won't do anymore, not showing the filter zone on rviz2 as it used to do.  

```
root@debix:/home/debix#ros2 launch nav2_costmap_filters_demo costmap_filter_info.launch.py 
[INFO] [launch]: All log files can be found below /root/.ros/log/2024-06-26-10-50-41-854739-debix-6472
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [lifecycle_manager-1]: process started with pid [6473]
[INFO] [map_server-2]: process started with pid [6475]
[INFO] [costmap_filter_info_server-3]: process started with pid [6477]
[costmap_filter_info_server-3] [INFO] [1719391841.956288667] [costmap_filter_info_server]: This is costmap filter info publisher
[map_server-2] [INFO] [1719391842.010769252] [filter_mask_server]:
[map_server-2]  filter_mask_server lifecycle node launched.
[map_server-2]  Waiting on external lifecycle transitions to activate
[map_server-2]  See https://design.ros2.org/articles/node_lifecycle.html for more information.
[map_server-2] [INFO] [1719391842.011617853] [filter_mask_server]: Creating
[lifecycle_manager-1] [INFO] [1719391842.343055765] [lifecycle_manager_costmap_filters]: Creating
[lifecycle_manager-1] [INFO] [1719391842.350849499] [lifecycle_manager_costmap_filters]: Creating and initializing lifecycle service clients
[costmap_filter_info_server-3] [INFO] [1719391842.399321731] [costmap_filter_info_server]:
[costmap_filter_info_server-3]  costmap_filter_info_server lifecycle node launched.
[costmap_filter_info_server-3]  Waiting on external lifecycle transitions to activate
[costmap_filter_info_server-3]  See https://design.ros2.org/articles/node_lifecycle.html for more information.
[lifecycle_manager-1] [INFO] [1719391842.997990058] [lifecycle_manager_costmap_filters]: Starting managed nodes bringup...
[lifecycle_manager-1] [INFO] [1719391842.998071979] [lifecycle_manager_costmap_filters]: Configuring filter_mask_server
[map_server-2] [INFO] [1719391842.998529652] [filter_mask_server]: Configuring
[map_server-2] [INFO] [map_io]: Loading yaml file: /home/Maps/Current/GeneralCostmap/ActiveKeepoutMap.yaml
[map_server-2] [DEBUG] [map_io]: resolution: 0.03
[map_server-2] [DEBUG] [map_io]: origin[0]: 0
[map_server-2] [DEBUG] [map_io]: origin[1]: 0
[map_server-2] [DEBUG] [map_io]: origin[2]: 0
[map_server-2] [DEBUG] [map_io]: free_thresh: 0.25
[map_server-2] [DEBUG] [map_io]: occupied_thresh: 0.65
[map_server-2] [DEBUG] [map_io]: mode: trinary
[map_server-2] [DEBUG] [map_io]: negate: 0
[map_server-2] [INFO] [map_io]: Loading image_file: /home/Maps/Current/GeneralCostmap/ActiveKeepoutMap.pgm
[map_server-2] [DEBUG] [map_io]: Read map /home/Maps/Current/GeneralCostmap/ActiveKeepoutMap.pgm: 431 X 252 map @ 0.03 m/cell
[lifecycle_manager-1] [INFO] [1719391843.027790948] [lifecycle_manager_costmap_filters]: Configuring costmap_filter_info_server
[costmap_filter_info_server-3] [INFO] [1719391843.028133159] [costmap_filter_info_server]: Configuring
[lifecycle_manager-1] [INFO] [1719391843.035168056] [lifecycle_manager_costmap_filters]: Activating filter_mask_server
[map_server-2] [INFO] [1719391843.035420088] [filter_mask_server]: Activating
[map_server-2] [INFO] [1719391843.035788175] [filter_mask_server]: Creating bond (filter_mask_server) to lifecycle manager.
[lifecycle_manager-1] [INFO] [1719391843.147892822] [lifecycle_manager_costmap_filters]: Server filter_mask_server connected with bond.
[lifecycle_manager-1] [INFO] [1719391843.147963052] [lifecycle_manager_costmap_filters]: Activating costmap_filter_info_server
[costmap_filter_info_server-3] [INFO] [1719391843.148337492] [costmap_filter_info_server]: Activating
[costmap_filter_info_server-3] [INFO] [1719391843.148439048] [costmap_filter_info_server]: Creating bond (costmap_filter_info_server) to lifecycle manager.
[lifecycle_manager-1] [INFO] [1719391843.258075229] [lifecycle_manager_costmap_filters]: Server costmap_filter_info_server connected with bond.
[lifecycle_manager-1] [INFO] [1719391843.258137952] [lifecycle_manager_costmap_filters]: Managed nodes are active
[lifecycle_manager-1] [INFO] [1719391843.258156881] [lifecycle_manager_costmap_filters]: Creating bond timer...
```

Do you know what could it be?
